lib/unison-sqlite/src/Unison/Sqlite/Transaction.hs

```diff
@@ -72,33 +81,110 @@ newtype Transaction a
+{-# SPECIALIZE runReadOnlyTransaction :: Connection -> ((forall x. Transaction x -> IO x) -> IO a) -> IO a #-}
+
+runReadOnlyTransaction_ :: Connection -> IO a -> IO a
+runReadOnlyTransaction_ conn action = do
+  bracketOnError_
+    (Connection.begin conn)
+    (ignoringExceptions (Connection.rollback conn))
+    ( do
+        result <- action
+        Connection.commit conn
+        pure result
+    )
+
+-- | Run a transaction that is known to perform at least one write.
+--
+-- The action is provided a function that peels off the 'Transaction' newtype without sending the corresponding
+-- BEGIN/COMMIT statements.
+--
+-- The transaction is never retried, so it is (more) safe to interleave arbitrary IO actions.
+runWriteTransaction :: MonadUnliftIO m => Connection -> ((forall x. Transaction x -> m x) -> m a) -> m a
+runWriteTransaction conn f =
+  withRunInIO \runInIO ->
+    uninterruptibleMask \restore ->
+      runWriteTransaction_
+        restore
+        100_000
+        conn
+        (runInIO (f (\transaction -> liftIO (unsafeUnTransaction transaction conn))))
+{-# SPECIALIZE runWriteTransaction :: Connection -> ((forall x. Transaction x -> IO x) -> IO a) -> IO a #-}
+
+runWriteTransaction_ :: (forall x. IO x -> IO x) -> Int -> Connection -> IO a -> IO a
+runWriteTransaction_ restore microseconds conn transaction = do
+  keepTryingToBeginImmediate restore conn microseconds
+  result <- restore transaction `onException` ignoringExceptions (Connection.rollback conn)
+  Connection.commit conn
+  pure result
+
+-- @BEGIN IMMEDIATE@ until success.
+keepTryingToBeginImmediate :: (forall x. IO x -> IO x) -> Connection -> Int -> IO ()
+keepTryingToBeginImmediate restore conn =
+  let loop microseconds =
+        try @_ @SqliteQueryException (Connection.beginImmediate conn) >>= \case
+          Left SqliteBusyException -> do
+            restore (threadDelay microseconds)
+            loop (microseconds * 2)
+          Left exception -> throwIO exception
+          Right () -> pure ()
+   in loop
+
+ignoringExceptions :: IO () -> IO ()
+ignoringExceptions action =
+  action `catchAny` \_ -> pure ()
+
+-- | Unwrap the transaction newtype, throwing away the sending of BEGIN/COMMIT + automatic retry.
+unsafeUnTransaction :: Transaction a -> Connection -> IO a
+unsafeUnTransaction (Transaction action) =
+  action
+
+-- | Perform an atomic sub-computation within a transaction; if it returns 'Left', it's rolled back.
+savepoint :: Transaction (Either a a) -> Transaction a
+savepoint (Transaction action) = do
+  Transaction \conn -> do
+    -- Generate a random name for the savepoint, so the caller isn't burdened with coming up with a name. Seems
+    -- extremely unlikely for this to go wrong (i.e. some super nested withSavepoint call that ends up generating the
+    -- same savepoint name twice in a single scope).
+    name <- Text.pack <$> replicateM 10 (Random.randomRIO ('a', 'z'))
```

Comment:
What do you think of accepting a name from the caller and suffixing it with a random string so that any savepoint errors will contain a human-readable name? I'm not sure it's necessary, but sqlite errors aren't always the nicest to read so it might help with debugging.

Reply:
could delete this code for now